File: backend-M2/src/main/java/com/fincore/BankingManagement/LoanOrigination/entity/service/LoanOriginationService.java
package com.fincore.BankingManagement.LoanOrigination.entity.service;

import com.fincore.BankingManagement.CreditCheck.entity.Customer;
import com.fincore.BankingManagement.LoanOrigination.entity.ApplicationStatus;
import com.fincore.BankingManagement.LoanOrigination.entity.LoanOrigination;
import com.fincore.BankingManagement.LoanOrigination.entity.Repository.CustomerRepo;
import com.fincore.BankingManagement.LoanOrigination.entity.dto.LoanApplicationRequest;
import com.fincore.BankingManagement.LoanOrigination.entity.dto.LoanApplicationResponse;
import com.fincore.BankingManagement.LoanOrigination.entity.Repository.LoanOriginationRepository;
import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class LoanOriginationService {

        private final LoanOriginationRepository repository;

        private final LoanProductService loanProductService;

        private final CustomerRepo customerRepository;

        @Transactional
        public LoanOrigination createLoanApplication(
                        LoanApplicationRequest request) {

                // =====================================================
                // 1. Validate Loan Type
                // =====================================================

                if (request.getLoanType() == null) {
                        throw new IllegalArgumentException(
                                        "Loan type is required");
                }

                // =====================================================
                // 2. Validate Loan Amount
                // =====================================================

                BigDecimal loanAmount = request.getLoanAmount();
                if (loanAmount == null) {
                        loanAmount = request.getRequestedAmount();
                }

                if (loanAmount == null ||
                                loanAmount.compareTo(BigDecimal.ZERO) <= 0) {

                        throw new IllegalArgumentException(
                                        "Loan amount must be greater than zero");
                }

                // =====================================================
                // 3. Validate Tenure
                // =====================================================

                if (request.getTenureMonths() == null ||
                                request.getTenureMonths() <= 0) {

                        throw new IllegalArgumentException(
                                        "Tenure must be greater than zero");
                }

                // =====================================================
                // 4. Get Customer Information (if customerId provided)
                // =====================================================

                Long customerId = request.getCustomerId();
                String customerName = request.getCustomerName();

                if (customerName == null) {
                        customerName = request.getFullName();
                }

                if (customerName == null) {
                        customerName = "Unknown Customer";
                }

                // Try to find customer if customerId is provided
                if (customerId != null) {
                        try {
                                Customer customer = customerRepository.findById(
                                                Math.toIntExact(customerId)).orElse(null);

                                if (customer != null) {
                                        customerName = customer.getFullName();
                                }
                        } catch (Exception e) {
                                // Customer not found, use provided name
                        }
                } else {
                        // Generate a default customer ID if not provided
                        customerId = 999999L; // Default ID
                }

                // =====================================================
                // 5. GET INTEREST RATE FROM LOAN PRODUCT OR REQUEST
                // =====================================================

                BigDecimal interestRate = request.getInterestRate();
                if (interestRate == null || interestRate.compareTo(BigDecimal.ZERO) <= 0) {
                        interestRate = loanProductService.getInterestRate(
                                        request.getLoanType());
                }

                // =====================================================
                // 6. CREATE LOAN ENTITY
                // =====================================================

                LoanOrigination loan = new LoanOrigination();

                // Set identifiers
                loan.setCustomerId(customerId);
                loan.setCustomerName(customerName);

                // Set personal information
                loan.setFullName(request.getFullName() != null ? request.getFullName() : customerName);

                if (request.getDateOfBirth() != null) {
                        try {
                                loan.setDateOfBirth(LocalDate.parse(request.getDateOfBirth()));
                        } catch (Exception e) {
                                // Invalid date format
                        }
                }

                loan.setGender(request.getGender());
                loan.setMobile(request.getMobile());
                loan.setEmail(request.getEmail());

                // Set address information
                loan.setAddress(request.getAddress());
                loan.setCity(request.getCity());
                loan.setState(request.getState());
                loan.setPincode(request.getPincode());

                // Set employment information
                loan.setEmploymentType(request.getEmploymentType());
                loan.setEmployerName(request.getEmployerName());
                loan.setJobTitle(request.getJobTitle());
                loan.setWorkExperience(request.getWorkExperience());

                // Set income information
                loan.setMonthlyIncome(request.getMonthlyIncome());
                loan.setOtherIncome(request.getOtherIncome());

                // Set loan information
                loan.setLoanType(request.getLoanType());
                loan.setLoanAmount(loanAmount);
                loan.setTenureMonths(request.getTenureMonths());
                loan.setInterestRate(interestRate);
                loan.setPurpose(request.getPurpose());

                // Set application status
                ApplicationStatus status = ApplicationStatus.PENDING;
                if (request.getApplicationStatus() != null) {
                        try {
                                status = ApplicationStatus.valueOf(
                                                request.getApplicationStatus().toUpperCase());
                        } catch (Exception e) {
                                status = ApplicationStatus.PENDING;
                        }
                }
                loan.setApplicationStatus(status);

                // Set application date
                loan.setApplicationDate(LocalDate.now());

                // =====================================================
                // 7. SAVE
                // =====================================================

                return repository.save(loan);
        }

        // =========================================================
        // GET ALL LOANS
        // =========================================================

        public List<LoanOrigination> getAllLoanApplications() {
                return repository.findAll();
        }

        // =========================================================
        // GET LOAN BY ID
        // =========================================================

        public LoanOrigination getLoanApplicationById(Long loanId) {
                return repository.findById(loanId)
                                .orElseThrow(() -> new RuntimeException(
                                                "Loan application not found: " + loanId));
        }

        // =========================================================
        // GET LOANS BY CUSTOMER
        // =========================================================

        public List<LoanOrigination> getLoansByCustomerId(Long customerId) {
                return repository.findByCustomerId(customerId);
        }

        // =========================================================
        // GET LOANS BY STATUS
        // =========================================================

        public List<LoanOrigination> getLoansByStatus(ApplicationStatus status) {
                return repository.findByApplicationStatus(status);
        }

        // =========================================================
        // UPDATE LOAN STATUS
        // =========================================================

        @Transactional
        public LoanOrigination updateLoanStatus(
                        Long loanId,
                        ApplicationStatus status) {

                LoanOrigination loan = getLoanApplicationById(loanId);
                loan.setApplicationStatus(status);
                return repository.save(loan);
        }

        // =========================================================
        // CONVERT ENTITY TO RESPONSE DTO
        // =========================================================

        public LoanApplicationResponse convertToResponse(LoanOrigination loan) {
                return LoanApplicationResponse.builder()
                                .loanId(loan.getLoanId())
                                .customerId(loan.getCustomerId())
                                .customerName(loan.getCustomerName())
                                .fullName(loan.getFullName())
                                .dateOfBirth(loan.getDateOfBirth())
                                .gender(loan.getGender())
                                .mobile(loan.getMobile())
                                .email(loan.getEmail())
                                .address(loan.getAddress())
                                .city(loan.getCity())
                                .state(loan.getState())
                                .pincode(loan.getPincode())
                                .employmentType(loan.getEmploymentType())
                                .employerName(loan.getEmployerName())
                                .jobTitle(loan.getJobTitle())
                                .workExperience(loan.getWorkExperience())
                                .monthlyIncome(loan.getMonthlyIncome())
                                .otherIncome(loan.getOtherIncome())
                                .loanType(loan.getLoanType())
                                .loanAmount(loan.getLoanAmount())
                                .tenureMonths(loan.getTenureMonths())
                                .interestRate(loan.getInterestRate())
                                .purpose(loan.getPurpose())
                                .applicationStatus(loan.getApplicationStatus())
                                .applicationDate(loan.getApplicationDate())
                                .createdAt(loan.getCreatedAt())
                                .updatedAt(loan.getUpdatedAt())
                                .build();
        }
        // CONVERT LIST OF ENTITIES TO RESPONSE DTOs

        public List<LoanApplicationResponse> convertToResponseList(List<LoanOrigination> loans) {
                return loans.stream()
                                .map(this::convertToResponse)
                                .collect(Collectors.toList());
        }
}

File: frontend/src/app/core/services/loan-origination.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of, BehaviorSubject } from 'rxjs';
import { catchError, map, tap } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export type LoanType =
  | 'Personal'
  | 'Home'
  | 'Vehicle'
  | 'Education'
  | 'Gold'
  | 'Other';

export type ApplicationStatus =
  | 'Pending'
  | 'Approved'
  | 'Rejected'
  | 'Draft'
  | 'Under Review'
  | 'Funded';

export interface LoanApplicationPayload {
  loanId?: number;

  // Customer / Applicant
  customerId?: number;
  fullName?: string;
  customerName?: string;
  dateOfBirth?: string;
  gender?: string;
  mobile?: string;
  email?: string;

  // Address
  address?: string;
  city?: string;
  state?: string;
  pincode?: string;

  // Employment
  employmentType?: string;
  employerName?: string;
  jobTitle?: string;
  workExperience?: string;

  // Income
  monthlyIncome?: number;
  otherIncome?: number;

  // Loan
  loanType: LoanType;
  loanAmount: number;
  requestedAmount?: number;
  tenureMonths: number;
  interestRate: number;
  purpose?: string;

  // Application
  applicationStatus?: ApplicationStatus;
  applicationDate?: string;
}

@Injectable({
  providedIn: 'root',
})
export class LoanOriginationService {
  private http = inject(HttpClient);

  private apiUrl = `${environment.apiUrl}/api/loan-origination`;

  private initialLoans: LoanApplicationPayload[] = [
    {
      loanId: 1001,
      customerId: 2001,
      customerName: 'Aditi Sharma',
      fullName: 'Aditi Sharma',
      loanType: 'Home',
      loanAmount: 3100000,
      requestedAmount: 3100000,
      tenureMonths: 240,
      interestRate: 7.35,
      purpose: 'Home purchase',
      applicationStatus: 'Pending',
      applicationDate: '2026-08-01',
    },

    {
      loanId: 1002,
      customerId: 2002,
      customerName: 'Rohan Mehta',
      fullName: 'Rohan Mehta',
      loanType: 'Personal',
      loanAmount: 2200000,
      requestedAmount: 2200000,
      tenureMonths: 120,
      interestRate: 8.1,
      purpose: 'Business expansion',
      applicationStatus: 'Approved',
      applicationDate: '2026-07-28',
    },

    {
      loanId: 1003,
      customerId: 2003,
      customerName: 'Sneha Patel',
      fullName: 'Sneha Patel',
      loanType: 'Vehicle',
      loanAmount: 850000,
      requestedAmount: 850000,
      tenureMonths: 60,
      interestRate: 9.5,
      purpose: 'Car purchase',
      applicationStatus: 'Pending',
      applicationDate: '2026-08-05',
    },
  ];

  private localLoans$ = new BehaviorSubject<LoanApplicationPayload[]>(
    this.initialLoans,
  );

  // =====================================================
  // CREATE LOAN APPLICATION
  // POST /api/loan-origination
  // =====================================================

  createLoanApplication(
    data: LoanApplicationPayload,
  ): Observable<LoanApplicationPayload> {
    const payload = {
      customerId: data.customerId,

      fullName: data.fullName,
      customerName: data.customerName || data.fullName,
      dateOfBirth: data.dateOfBirth,
      gender: data.gender,
      mobile: data.mobile,
      email: data.email,

      address: data.address,
      city: data.city,
      state: data.state,
      pincode: data.pincode,

      employmentType: data.employmentType,
      employerName: data.employerName,
      jobTitle: data.jobTitle,
      workExperience: data.workExperience,

      monthlyIncome: data.monthlyIncome,
      otherIncome: data.otherIncome,

      loanType: data.loanType,
      loanAmount: Number(data.loanAmount),
      requestedAmount: Number(data.requestedAmount ?? data.loanAmount),
      tenureMonths: Number(data.tenureMonths),
      interestRate: Number(data.interestRate),
      purpose: data.purpose || '',
      applicationStatus: data.applicationStatus || 'Pending',
    };

    console.log('POST Loan Application:', payload);

    return this.http.post<LoanApplicationPayload>(this.apiUrl, payload).pipe(
      tap((created) => {
        this.addLocalRecord(created);
      }),
    );
  }

  // =====================================================
  // GET ALL LOAN APPLICATIONS
  // GET /api/loan-origination
  // =====================================================

  getAllLoanApplications(): Observable<LoanApplicationPayload[]> {
    return this.http.get<LoanApplicationPayload[]>(this.apiUrl).pipe(
      tap((res) => {
        if (res && res.length) {
          this.localLoans$.next(res);
        }
      }),

      catchError((error) => {
        console.error('Failed to get loan applications:', error);

        return of(this.localLoans$.value);
      }),
    );
  }

  // =====================================================
  // GET LOAN BY ID
  // GET /api/loan-origination/{loanId}
  // =====================================================

  getLoanApplicationById(
    loanId: number,
  ): Observable<LoanApplicationPayload | null> {
    return this.http
      .get<LoanApplicationPayload>(`${this.apiUrl}/${loanId}`)
      .pipe(
        catchError((error) => {
          console.error('Failed to get loan:', error);

          const found = this.localLoans$.value.find(
            (loan) => loan.loanId === Number(loanId),
          );

          return of(found || null);
        }),
      );
  }

  // =====================================================
  // GET LOANS BY CUSTOMER ID
  // GET /api/loan-origination/customer/{customerId}
  // =====================================================

  getLoansByCustomerId(
    customerId: number,
  ): Observable<LoanApplicationPayload[]> {
    return this.http
      .get<LoanApplicationPayload[]>(`${this.apiUrl}/customer/${customerId}`)
      .pipe(
        catchError((error) => {
          console.error('Failed to get customer loans:', error);

          const found = this.localLoans$.value.filter(
            (loan) => loan.customerId === Number(customerId),
          );

          return of(found);
        }),
      );
  }

  // =====================================================
  // GET LOANS BY STATUS
  // GET /api/loan-origination/status/{status}
  // =====================================================

  getLoansByStatus(
    status: ApplicationStatus,
  ): Observable<LoanApplicationPayload[]> {
    return this.http
      .get<
        LoanApplicationPayload[]
      >(`${this.apiUrl}/status/${encodeURIComponent(status)}`)
      .pipe(
        catchError((error) => {
          console.error('Failed to get loans by status:', error);

          const found = this.localLoans$.value.filter(
            (loan) => loan.applicationStatus === status,
          );

          return of(found);
        }),
      );
  }

  // =====================================================
  // UPDATE LOAN STATUS
  // PUT /api/loan-origination/{loanId}/status
  // =====================================================

  updateLoanStatus(
    loanId: number,
    status: ApplicationStatus,
  ): Observable<LoanApplicationPayload> {
    return this.http
      .put<LoanApplicationPayload>(
        `${this.apiUrl}/${loanId}/status?status=${encodeURIComponent(status)}`,
        {},
      )
      .pipe(
        tap((updated) => {
          this.updateLocalStatus(loanId, status);
        }),

        catchError((error) => {
          console.error('Failed to update loan status:', error);

          throw error;
        }),
      );
  }

  // =====================================================
  // LOCAL CACHE
  // =====================================================

  private addLocalRecord(record: LoanApplicationPayload): void {
    const current = this.localLoans$.value;

    const withoutExisting = current.filter((r) => r.loanId !== record.loanId);

    this.localLoans$.next([record, ...withoutExisting]);
  }

  private updateLocalStatus(loanId: number, status: ApplicationStatus): void {
    const updated = this.localLoans$.value.map((loan) =>
      loan.loanId === Number(loanId)
        ? {
            ...loan,
            applicationStatus: status,
          }
        : loan,
    );

    this.localLoans$.next(updated);
  }
}
